### DISCUSSION TITLE
AutoDisposeFutureProvider was disposed during loading state, yet no value could be emitted

### DISCUSSION THREAD

**Discussion by qwadrox** [rrousselGit/riverpod#3223] [Q&A] [answered]
Could you enlighten me in which specific scenario this error would occur?  

It causes me a headache since, when this occurs, it is mapped to the 'when's error function -> AsyncValueX, and this results in displaying an irrelevant error widget for some time.

```bash
Bad state: The provider AutoDisposeFutureProvider<GetHomeResponseDto?>#f72d1(UserGuidRequestDto(userGuid: something)) was disposed during loading state, yet no value could be emitted.
StackTrace: #0      FutureHandlerProviderElementMixin.dispose (package:riverpod/src/async_notifier/base.dart:460)
#1      ProviderContainer._disposeProvider (package:riverpod/src/framework/container.dart:336)
#2      ProviderScheduler._performDispose (package:riverpod/src/framework/scheduler.dart:128)
#3      ProviderScheduler._task (package:riverpod/src/framework/scheduler.dart:83)
#4      ProviderScheduler.vsync.<anonymous closure>.invoke (package:riverpod/src/framework/scheduler.dart:41)
#5      _UncontrolledProviderScopeElement.build (package:flutter_riverpod/src/framework.dart:390)
#6      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5471)
#7      Element.rebuild (package:flutter/src/widgets/framework.dart:5187)
#8      BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2895)
#9      WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:984)
#10     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:457)
#11     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1325)
#12     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1255)
#13     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1113)
#14     _invoke (dart:ui/hooks.dart:312)
#15     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:383)
#16     _drawFrame (dart:ui/hooks.dart:283)
```

The best thing would be to ignore this cause seemingly it doesn't break anything and I encountered with it when app was paused and resumed.

**Comment by Maatteogekko** [comment]
The error comes from here:
https://github.com/rrousselGit/riverpod/blob/655762024f9927c5d7c791fbbb9f92a7183ef53a/packages/riverpod/lib/src/async_notifier/base.dart#L457-L461

It seems harmless. What confuses me is why disposing while loading is considered an error.

To give a practical example, in many of my AsyncNotifiers I do something like the following early in the build method
```dart
if(!someCondition) return future;
```

If the condition never resolves to true and the notifier gets disposed, which I expect to happen normally during the various lifecycles of the app, the error is thrown.

---

On the discord a similar question has already been answered: https://discord.com/channels/765557403865186374/765557404766830614/1093429859532943391
tl;dr: to avoid the error an initial state can be used, instead of leaving the notifier on loading.

Throwing an error to indicate a _possible_ mis-design seems a bit too drastic though. @rrousselGit wdyt?

**Reply by rrousselGit** [reply, marked as answer]
Look at it this way: What else could we return here?
Having the future never resolve is undesirable. That's both a memory leak and a possible infinite spinner.


You should instead ponder why your provider got disposed when it was still in use. Seems like you're missing a ref.watch/listen somehow.

Doing`await ref.read(provider.future)` isn't safe.

**Reply by rrousselGit** [reply]
Note that this is also consistent with how `Stream.first` works. If the stream is closed before a value is emitted, `.first` will throw

**Comment by scopendo** [comment]
If `await ref.read(provider.future)` isn't safe, then does it follow that neither is:

```dart
RefreshIndicator(
  onRefresh: () => ref.refresh(provider.future),
)
```
